src/vision/status_quo/alan_thinks_he_needs_async_locks.md

```diff
@@ -0,0 +1,121 @@
+    |                     ---- required by this bound in `tokio::spawn`
+
+note: future is not `Send` as this value is used across an await
+   --> src/handler.rs:787:9
+      |
+786   |         let routes = self.state.lock().unwrap();
+      |             - has type `std::sync::MutexGuard<'_, HashMap<String, Url>>` which is not `Send`
+787   |         Ok(self.client.perform_request(route, request).await?)
+      |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ await occurs here, with `routes` maybe used later
+788   |     })
+      |     - `routes` is later dropped here
+```
+
+Alan stops and takes a deep breath. He tries his best to make sense of the error message. He sort of understands the issue the compiler is telling him. Apparently `routes` is not marked as `Send`, and because it is still alive over a call to `await`, it is making the future his handler returns not `Send`. And tokio's `spawn` function seems to require that the future it received be `Send`. 
+
+Alan reaches the boundaries of his knowledge of Rust, so he reaches out over chat to ask his co-worker Barbara for help. Not wanting to bother her, Alan provides the context he's already figured out for himself.
+
+Barbara knows that mutex guards are not `Send` because sending mutex guards to different threads is not a good idea. She suggests looking into async locks which can be held across await points because they are `Send`. Alan looks into the tokio documentation for more info and is easily able to move the use of the standard library's mutex to tokio's mutex. It compiles!
+
+Alan ships his code and it gets a lot of usage. After a while, Alan notices some potential performance issues. It seems his proxy handler does not have the throughput he would expect. Barbara, having newly joined his team, sits down with him to take a look at potential issues. Barbara is immediately worried by the fact that the lock is being held much longer than it needs to be. The lock only needs to be held while accessing the route and not during the entire duration of the downstream request.
+
+She suggests to Alan to switch to not holding the lock across the I/O operations. Alan first tries to do this by explicitly cloning the url and dropping the lock before the proxy request is made:
+
+```rust
+impl Proxy {
+  async fn handle(&self, key: String, request: Request) -> crate::Result<Response> {
+      let routes = self.state.lock().unwrap();
+      let route = routes.get(key).unwrap_or_else(crate::error::MissingRoute)?.clone();
+      drop(routes);
+      Ok(self.client.perform_request(route, request).await?)
+  }
+}
+```
+
+Unfortunately this does not seem to fix the issue - he gets the same error message as before. Confused Alan goes to Barbara for advice. She is also confused, and it takes several minutes of exploration before she comes to a solution that works - wrapping the mutex access in a block and implicitly dropping the mutex:
```

Comment:
This seems out of date -- he's not getting an error, right? He's getting a slowdown?

Unfortunately, the bug that is causing the `{...}` to be required *does* affect compilation errors but probably *wouldn't* affect performance issues (basically the compiler overapproximates how long something is live) so I think some restructuring of the story is needed here (e.g., he can try adding the `drop` before adopting the tokio lock)